client/logs.go

```diff
@@ -38,10 +41,9 @@ type LogsOptions struct {
 	// all services).
 	Services []string
 
-	// Total number of logs to fetch (before following if calling FollowLogs).
-	// If nil, use Pebble's default (10). If negative, fetch all buffered logs.
-	// If set to zero when calling FollowLogs, write no logs before following.
-	N *int
+	// Number of logs to fetch (before following if calling FollowLogs). To
+	// fetch all buffered logs, set N to AllLogs.
+	N int
```

Comment:
This is solving part of the problem by removing the pointer usage and making it a bit more clear, but it's still missing part of the problem: the current _server_ API is confusing. The default value is being set by the command, when it should really be in the server when N is unset. Otherwise, what sense it makes to request logs, and then get neither what's buffered nor what is coming after (follow)? It's an awkward reply, and we're addressing that not in the server nor in the client, but way down in the command. I think we should really fix the server.

My suggestion in the issue was to have the default behavior when N is zero being a nice outcome in both cases: if following, just follow and ignore the buffer. If not following, show a default number server-defined. To show all logs, we don't need any special logic or constant, since the client can trivially provide a large number as desired

To keep compatibility, we can allow the server to support -1 specifically, but it needs to be clear about it. Right now that whole side of the logic looks somewhat accidental instead of intentional when looking at the API code

Reply:
Ah, I'd originally misinterpreted your comment [here](https://github.com/canonical/pebble/issues/83#issuecomment-973980643) as saying we shouldn't change the server API, but I see now I was misreading it. Well, there are a few aspects to this:

1) Do we need a special case for "all"? I think we do. For one, it's a more direct API than specifying n=1000000 or whatever, and the user of the API won't have to think "wait, *how* big do I have to make it?" or worse, choose something like n=10000 that's not necessarily big enough. Also, the `RingBuffer` implementation and API needs to know, because it calls `TailIterator()` if n<0, or `HeadIterator(n)` if n>=0. In the latter case, `HeadIterator` scans back the given number of lines byte-by-byte, so it'd be quite inefficient to scan through the whole buffer if you just want it all.

2) Should we change the server API? I don't think it's particularly awkward -- it doesn't involve pointers as it's just a query parameter. Not specifying the `n` query parameter means "use the server default" in both follow and non-follow cases, which makes sense to me. The only special case is negative `n`, which means "all". We could change that to literally `?n=all` (we use `-n=all` in the CLI already).

3) I think it's confusing to have the Go client have a default of 10 in one case (non-follow) but a default of 0 in the other (follow). In the `pebble logs` CLI command, we want `pebble logs` and `pebble logs -f` to have the same default, like `tail` and `tail -f` both show the last 10 buffered logs. But with your suggested Go client change, we'd have to have a client-side default for `pebble logs -f` anyway.

4) Partly because of the above, it seems to me it was a bad idea to have a server-specified default at all. That should be something that's client-dependent. For example, on the CLI a default of 10 is good, but in a web UI, maybe 100 would make more sense. That's why I got rid of the default in the Go client in this PR -- I think any default should be client-specified (essentially making `LogsOptions.N` a required field).

My concrete suggestion is we keep the server API as is (or maybe tweak it to allow `n=all` as well as `n=-1`), and make the Go client change suggested in this PR. Alternatively for the Go client, we could have

```go
type LogsOptions struct {
   ...
   Fetch LogsFetch // one of FetchN (zero value), FetchDefault, FetchAll
   N     int       // not used if Fetch is FetchDefault or FetchAll 
}
```

If I'm not explaining myself clearly in the above, let's chat on video.

Reply:
Okay, I think that's reasonable, thanks. API design is hard! :-) I've made this change to the Go client and the same tweak to the server API and updated the PR -- ready for re-review.